File: Inkwell/Rendering/ConstellationClient.swift
//
//  ConstellationClient.swift
//  Inkwell
//
//  Queries the microcosm.blue Constellation API — a global AT Protocol
//  backlink index — to discover records (comments, quotes, mentions)
//  from ANY repository that reference a given subject URI.
//
//  Public instance: https://constellation.microcosm.blue
//

import Foundation
import OSLog

// MARK: - ConstellationClient

/// Queries the Constellation backlink index for cross-repo record discovery.
///
/// Without Constellation (or a similar backend), AT Protocol records stored
/// in other users' repositories are undiscoverable — each user's records are
/// siloed in their own PDS. Constellation indexes the firehose so you can ask
/// "which records across the entire network link to URI X?"
///
/// This client only does *discovery* — it returns backlink metadata (DID,
/// collection, rkey). To get the actual record content, hydrate each backlink
/// via `com.atproto.repo.getRecord` on the commenter's PDS.
enum ConstellationClient {
    private static let logger = Logger(subsystem: "uk.ewancroft.Inkwell", category: "Constellation")
    private static let baseURL = "https://constellation.microcosm.blue"

    private static let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        config.httpAdditionalHeaders = [
            "Accept": "application/json",
            "User-Agent": "Inkwell/1.0",
        ]
        return URLSession(configuration: config)
    }()

    // MARK: - Backlink Types

    /// A single backlink from Constellation's `getBacklinks` response.
    struct Backlink: Decodable, Sendable {
        let did: String
        let collection: String
        let rkey: String

        /// The full AT-URI of the linking record, constructed from its parts.
        var recordURI: String {
            "at://\(did)/\(collection)/\(rkey)"
        }
    }

    /// Response wrapper for `getBacklinks`.
    private struct GetBacklinksResponse: Decodable {
        let records: [Backlink]?
        let cursor: String?
    }

    // MARK: - Public API

    /// Finds all records (across the entire AT Protocol network) that link to
    /// the given subject via the given source collection + path.
    ///
    /// - Parameters:
    ///   - subject: The target AT-URI, DID, or HTTPS URL.
    ///   - source: The source collection and JSON path, joined by `:`.
    ///     e.g. `"pub.leaflet.comment:subject"` for comments on a document.
    ///   - limit: Maximum backlinks per page (≤ 100).
    ///   - cursor: Pagination cursor from a previous page.
    /// - Returns: Backlinks and an optional cursor for the next page.
    static func getBacklinks(
        subject: String,
        source: String,
        limit: Int = 50,
        cursor: String? = nil
    ) async throws -> (backlinks: [Backlink], cursor: String?) {
        guard var components = URLComponents(string: "\(baseURL)/xrpc/blue.microcosm.links.getBacklinks") else {
            throw URLError(.badURL)
        }

        var queryItems = [
            URLQueryItem(name: "subject", value: subject),
            URLQueryItem(name: "source", value: source),
            URLQueryItem(name: "limit", value: String(min(limit, 100))),
        ]
        if let cursor {
            queryItems.append(URLQueryItem(name: "cursor", value: cursor))
        }
        components.queryItems = queryItems

        guard let url = components.url else {
            throw URLError(.badURL)
        }

        logger.debug("[Constellation] getBacklinks subject=\(subject) source=\(source)")

        let (data, response) = try await session.data(from: url)

        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        guard (200...299).contains(http.statusCode) else {
            logger.error("[Constellation] HTTP \(http.statusCode) for \(url.absoluteString)")
            throw URLError(.badServerResponse)
        }

        let decoded = try JSONDecoder().decode(GetBacklinksResponse.self, from: data)
        let backlinks = decoded.records ?? []
        logger.info("[Constellation] getBacklinks returned \(backlinks.count) backlinks, cursor=\(decoded.cursor ?? "nil")")
        return (backlinks, decoded.cursor)
    }

    /// Finds all comment backlinks (`pub.leaflet.comment`) referencing a
    /// standard.site document AT-URI.
    ///
    /// Paginates through all results automatically.
    static func getCommentBacklinks(
        documentURI: String,
        maximumCount: Int = 200
    ) async -> [Backlink] {
        await paginateBacklinks(
            subject: documentURI,
            source: "pub.leaflet.comment:subject",
            maximumCount: maximumCount
        )
    }

    /// Finds all recommend backlinks (`site.standard.graph.recommend`)
    /// referencing a document AT-URI.
    static func getRecommendBacklinks(
        documentURI: String,
        maximumCount: Int = 200
    ) async -> [Backlink] {
        await paginateBacklinks(
            subject: documentURI,
            source: "site.standard.graph.recommend:document",
            maximumCount: maximumCount
        )
    }

    /// Finds Bluesky posts that link to a URL via facets or external embeds.
    ///
    /// This mirrors leaflet.pub's `getConstellationBacklinks()` which queries
    /// two source paths to catch both link-in-text and link-card embeds.
    static func getDocumentMentionBacklinks(
        url: String,
        maximumCount: Int = 200
    ) async -> [Backlink] {
        let sourceFacets = "app.bsky.feed.post:facets[].features[app.bsky.richtext.facet#link].uri"
        let sourceEmbeds = "app.bsky.feed.post:embed.external.uri"

        async let facetLinks = paginateBacklinks(
            subject: url,
            source: sourceFacets,
            maximumCount: maximumCount
        )
        async let embedLinks = paginateBacklinks(
            subject: url,
            source: sourceEmbeds,
            maximumCount: maximumCount
        )

        let (facets, embeds) = await (facetLinks, embedLinks)
        var seen = Set<String>()
        var merged: [Backlink] = []
        for link in facets + embeds {
            let key = "\(link.did):\(link.rkey)"
            if seen.insert(key).inserted {
                merged.append(link)
            }
        }
        return merged
    }

    // MARK: - Private

    /// Paginates through `getBacklinks` results until exhausted or the limit
    /// is reached.
    private static func paginateBacklinks(
        subject: String,
        source: String,
        maximumCount: Int
    ) async -> [Backlink] {
        var all: [Backlink] = []
        var cursor: String?

        repeat {
            guard let result = try? await getBacklinks(
                subject: subject,
                source: source,
                limit: 50,
                cursor: cursor
            ) else { break }

            all.append(contentsOf: result.backlinks)
            cursor = result.cursor
        } while cursor != nil && all.count < maximumCount

        return Array(all.prefix(maximumCount))
    }
}

File: Android/app/src/main/java/uk/ewancroft/inkwell/data/remote/ConstellationClient.kt
/**
 * Queries the microcosm.blue Constellation API — a global AT Protocol backlink
 * index — for cross-repo discovery of comments, recommends, and mentions.
 *
 * Without this, records in other users' repositories are undiscoverable.
 * Constellation indexes the full AT Protocol firehose so we can ask
 * "which records across the entire network link to URI X?"
 *
 * Mirrors Inkwell iOS ConstellationService: same endpoints, same pagination
 * strategy, same convenience method split (comment / recommend / mention).
 */
package uk.ewancroft.inkwell.data.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.decodeFromJsonElement
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import uk.ewancroft.inkwell.shared.graph.CollectionNsids
import uk.ewancroft.inkwell.shared.xrpc.XrpcEndpoints
import java.util.concurrent.TimeUnit
import uk.ewancroft.inkwell.data.model.bluesky.ConstellationBacklink
import uk.ewancroft.inkwell.data.model.bluesky.ConstellationResponse
import uk.ewancroft.inkwell.shared.constellation.ConstellationPagination
import uk.ewancroft.inkwell.shared.constellation.ConstellationBacklink as SharedBacklink
import uk.ewancroft.inkwell.shared.constellation.ConstellationResponse as SharedResponse
import uk.ewancroft.inkwell.shared.constellation.ConstellationSourcePaths
import uk.ewancroft.inkwell.shared.validation.JsonSafety

object ConstellationClient {
    private val json = Json { ignoreUnknownKeys = true }

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    // ── Backlink Query ───────────────────────────────────────────────────

    /** Finds all records that link to the given subject via the given source. */
    suspend fun getBacklinks(
        subject: String,
        source: String,
        limit: Int = 50,
        cursor: String? = null
    ): ConstellationResponse = withContext(Dispatchers.IO) {
        val url = "${XrpcEndpoints.CONSTELLATION_API}${XrpcEndpoints.MICROCOSM_GET_BACKLINKS}"
            .toHttpUrlOrNull()
            ?.newBuilder()
            ?.addQueryParameter("subject", subject)
            ?.addQueryParameter("source", source)
            ?.addQueryParameter("limit", limit.toString())
            ?.apply { cursor?.let { addQueryParameter("cursor", it) } }
            ?.build()
            ?: throw java.io.IOException("Invalid Constellation endpoint")

        val request = Request.Builder().url(url).get().build()
        client.newCall(request).execute().use { response ->
            val bodyString = response.body?.readBoundedUtf8()
            if (!response.isSuccessful || bodyString == null) {
                throw java.io.IOException(
                    "Constellation getBacklinks failed: HTTP ${response.code}"
                )
            }
            val element = json.parseToJsonElement(bodyString)
            check(JsonSafety.isSafe(element)) { "Constellation response exceeded structural safety limits" }
            json.decodeFromJsonElement<ConstellationResponse>(element)
        }
    }

    // ── Pagination ───────────────────────────────────────────────────────

    /** Paginates through all backlink results. */
    suspend fun paginateBacklinks(
        subject: String,
        source: String,
        maxCount: Int = 200
    ): List<ConstellationBacklink> {
        val shared = ConstellationPagination.paginateBacklinks(
            fetchPage = { limit, cursor -> getBacklinks(subject, source, limit, cursor).toShared() },
            maxCount = maxCount
        )
        return shared.map { it.toAndroid() }
    }

    /**
     * Total recommend count for a document, across the whole network.
     *
     * Mirrors Inkwell iOS `fetchRecommendCount(for:)`: request a single
     * record first: if there's no next cursor the whole result set fit in
     * that page, so its size is the count. Otherwise fall back to full
     * pagination via [getRecommendBacklinks].
     */
    suspend fun getRecommendCount(documentUri: String): Int {
        val first = try {
            getBacklinks(
                subject = documentUri,
                source = "${CollectionNsids.GRAPH_RECOMMEND}:document",
                limit = 1
            )
        } catch (_: Exception) {
            return 0
        }
        if (first.cursor == null) return first.records.orEmpty().size
        return getRecommendBacklinks(documentUri).size
    }

    /** Recommends (standard.site graph edges) pointing at this document. */
    suspend fun getRecommendBacklinks(documentUri: String): List<ConstellationBacklink> =
        paginateBacklinks(documentUri, "${CollectionNsids.GRAPH_RECOMMEND}:document")

    /** Comments (Leaflet pub.leaflet.comment records) pointing at this document. */
    suspend fun getCommentBacklinks(documentUri: String): List<ConstellationBacklink> =
        paginateBacklinks(documentUri, "${CollectionNsids.LEAFLET_COMMENT}:subject")

    /**
     * Mentions in Bluesky posts: searches both link facets and embed.external URIs.
     * Deduplicates by (did, rkey) since a single post could have both a facet
     * link and an embed URI referencing the same document.
     */
    suspend fun getDocumentMentionBacklinks(url: String): List<ConstellationBacklink> = coroutineScope {
        val facets = async { paginateBacklinks(
            url, ConstellationSourcePaths.MENTION_FACET_LINK
        ) }
        val embeds = async { paginateBacklinks(
            url, ConstellationSourcePaths.EMBED_EXTERNAL_URI
        ) }
        val f = facets.await()
        val e = embeds.await()
        ConstellationPagination.deduplicate((f + e).map { it.toShared() }).map { it.toAndroid() }
    }
}

// ── Type Conversions ─────────────────────────────────────────────────────

private fun ConstellationBacklink.toShared(): SharedBacklink = SharedBacklink(
    did = did,
    collection = collection,
    rkey = rkey
)

private fun ConstellationResponse.toShared(): SharedResponse = SharedResponse(
    records = records.orEmpty().map { it.toShared() },
    cursor = cursor
)

private fun SharedBacklink.toAndroid(): ConstellationBacklink = ConstellationBacklink(
    did = did,
    collection = collection,
    rkey = rkey
)
